### ISSUE TITLE
ValueError: not allowed to raise maximum limit

### ISSUE THREAD

**Issue by dmelcer9** [bigcode-project/bigcodebench#8]
I am trying to do some evaluations of bigcodebench with non-greedy decoding, and I am running into an error:

Steps to reproduce:

1. python version 3.9.19, CUDA version 12.0, Docker version 20.10.23, RAM > 128GB
2. `bigcodebench.generate --model bigcode/starcoder2-15b --subset complete --bs 4 --n_samples 12 --temperature 1 --backend vllm --save_path output.jsonl`
3. `bigcodebench.sanitize --samples output.jsonl --calibrate`
4. `docker run -it --entrypoint bigcodebench.evaluate -v $(pwd):/app bigcodebench/bigcodebench-evaluate:latest --samples output-sanitized-calibrated.jsonl --subset complete`

Full output of the last command:
```
Downloading dataset from https://github.com/bigcode-project/bigcodebench-annotation/releases/download/v0.1.0/BigCodeBench.jsonl.gz
100% [..........................................................................] 1127888 / 1127888
Asserting the groundtruth...
  0%|                                                                     | 0/1140 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "/usr/local/bin/bigcodebench.evaluate", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.10/site-packages/bigcodebench/evaluate.py", line 251, in main
    evaluate(args)
  File "/usr/local/lib/python3.10/site-packages/bigcodebench/evaluate.py", line 115, in evaluate
    expected_time = get_groundtruth(problems, dataset_hash, flags.check_gt_only)
  File "/usr/local/lib/python3.10/site-packages/bigcodebench/evaluate.py", line 52, in get_groundtruth
    expected_time[task_id] = trusted_exec(
  File "/usr/local/lib/python3.10/site-packages/bigcodebench/gen/util/__init__.py", line 29, in trusted_exec
    reliability_guard(maximum_memory_bytes=maximum_memory_bytes)
  File "/usr/local/lib/python3.10/site-packages/bigcodebench/eval/utils.py", line 295, in reliability_guard
    resource.setrlimit(
ValueError: not allowed to raise maximum limit
```

Troubleshooting steps tried so far:
- Running the last command with `sudo` does not help.
- With `--no-gt` just causes a whole bunch of these identical errors to show up, presumably one for each thread?

**Comment by terryyz** [comment]
Hi @dmelcer9, thanks for reporting the issue!

It appears that the problem is not related to the setup of the generation but more to the memory allocation. Given that you mentioned running the machine with more than 128GB of RAM, this is indeed a bit unusual.

I haven't encountered such issues before. Based on a [Stack Overflow discussion](https://stackoverflow.com/questions/10155301/python-valueerror-not-allowed-to-raise-maximum-limit), it might be possible that other processes on your machine are consuming RAM while you run the evaluation.

To investigate further, I recommend using `psutil.virtual_memory()` to inspect the RAM usage during the evaluation. This can provide insight into whether other processes are impacting the available memory.

Please note that the resource limit setup is the same as [EvalPlus](https://github.com/evalplus/evalplus/blob/96663fc1a08af9b1ad950c133a657c783110b5c3/evalplus/eval/utils.py#L124), except that ours has a much higher limit.

**Comment by dmelcer9** [comment]
Looking at htop, only about 8GB is being used for other processes, leaving well over 128GB free. I was curious if this was a `ulimit` issue, but it appears not:

```
$ ulimit -a
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 30446
max locked memory       (kbytes, -l) unlimited
max memory size         (kbytes, -m) unlimited
open files                      (-n) 8192
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 10240
cpu time               (seconds, -t) unlimited
max user processes              (-u) unlimited
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited
```

**Comment by terryyz** [comment]
Have you tried to modify the code and set it up with a lower limit? I guess the current limit could be too high for the data limit and stack limit.

I may add the pass arguments for the evaluation part, so you can adjust the limit yourself. The original purpose of setting a high limit is to pass the tasks involved with `tensorflow`.

**Comment by dmelcer9** [comment]
It seems the stack limit specifically would be the issue: 128GB is quite large as a stack size, it seems like the hard cap for stack size on my machine is ~10MB. Bigcodebench appears to leave the stack limit alone on Mac hosts as well- do the `tensorflow` tests succeed on that? I'd imagine that the heap may grow large, but not the stack unless the code was doing some quite deep recursion. Thank you for looking into this!

**Comment by terryyz** [comment]
Yeah, I think a good solution is to add optional passing arguments and decrease the limit for data and stack. I'll do a new version very shortly.

**Comment by terryyz** [comment]
Hi @dmelcer9, `v0.1.6` has been released. The docker images should be updated very soon. Please let me know if the new version works on your local machine :)

**Comment by terryyz** [comment]
Closed the issue for now. Feel free to reopen it if you have further questions!

**Comment by dmelcer9** [comment]
Seems that the docker image was only partially updated- I get a different error `/usr/local/lib/python3.10/site-packages/bigcodebench/evaluate.py:137: UserWarning: Task BigCodeBench/XXXX is found in the samples but not found in the dataset`, for all samples, but the line numbers don't actually match up with the where the warning is generated in `evaluate.py`. 

Running the docker version with `--check-gt-only` gives the following:

```
Downloading readme: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8.78k/8.78k [00:00<00:00, 40.0MB/s]
Downloading data: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2.39M/2.39M [00:00<00:00, 12.9MB/s]
Generating v0.1.0_hf split: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1140/1140 [00:00<00:00, 65621.44 examples/s]
Creating json from Arrow format:   0%|                                                                                                                                                                                                                                                         | 0/2 [00:00<?, ?ba/s]

Asserting the groundtruth...
0it [00:00, ?it/s]
Expected outputs computed in 0.00s
```

This doesn't seem to be a networking issue; isolating the docker container with `network="none"` produces a different error.

Testing the dockerless version with `--check-gt-only` seems to work fine.

**Comment by terryyz** [comment]
Hmm, I'll double-check the docker in a few hours. Thanks for noting this issue! Reopened it for now.

**Comment by terryyz** [comment]
I apologize for any inconvenience caused. I may accidentally pull the main branch inside the Docker environment before merging the v0.1.6 branch. I'm currently working on creating the new Docker images to resolve this issue. Should be finished soon.

**Comment by terryyz** [comment]
I released `v0.1.7` to enhance the evaluation pipeline. When I was fixing some other bugs, I noticed that the evaluate container may have some memory issues on my machine. I'm not sure if that's due to my own environment or the docker itself. Could you check it on your machine to see if it works?

Note that the Generate image is still being pushed.

**Comment by terryyz** [comment]
Closed the issue :)
